File: 10-factory-wrap/node-addon-api/src/addon.cc
#include <napi.h>
#include "my_object.h"

Napi::Object CreateObject(const Napi::CallbackInfo& info) {
  if (info.Length() <= 0 || !info[0].IsNumber()) {
    Napi::TypeError::New(info.Env(), "Number expected")
        .ThrowAsJavaScriptException();
    Napi::Object obj = Napi::Object::New(info.Env());
    return obj;
  }
  return MyObject::NewInstance(info.Env(), info[0]);
}

Napi::Object InitAll(Napi::Env env, Napi::Object exports) {
  Napi::Object new_exports =
      Napi::Function::New(env, CreateObject, "CreateObject");
  return MyObject::Init(env, new_exports);
}

NODE_API_MODULE(addon, InitAll)

File: 10-factory-wrap/node-addon-api/addon.js
let MyObject = require('bindings')('addon');

console.log(MyObject)

let obj = MyObject(10);
console.log(obj.plusOne()); // 11
console.log(obj.plusOne()); // 12
console.log(obj.plusOne()); // 13

let obj2 = MyObject(20);
console.log(obj2.plusOne()); // 21
console.log(obj2.plusOne()); // 22
console.log(obj2.plusOne()); // 23
